si5351/si5351.go

```
@@ -0,0 +1,443 @@
+
+	// Basic validation
+	if !d.initialised {
+		return errNotInitialised
+	}
+	// mult = 15..90
+	if !((mult > 14) && (mult < 91)) {
+		return errInvalidParameter
+	}
+	// Avoid divide by zero
+	if !(denom > 0) {
+		return errInvalidParameter
+	}
+	// 20-bit limit
+	if !(num <= 0xFFFFF) {
+		return errInvalidParameter
+	}
+	// 20-bit limit
+	if !(denom <= 0xFFFFF) {
+		return errInvalidParameter
+	}
+
+	// PLL Multiplier Equations
+	//
+	// P1 register is an 18-bit value using following formula:
+	//
+	// 	P1[17:0] = 128 * mult + floor(128*(num/denom)) - 512
+	//
+	// P2 register is a 20-bit value using the following formula:
+	//
+	// 	P2[19:0] = 128 * num - denom * floor(128*(num/denom))
+	//
+	// P3 register is a 20-bit value using the following formula:
+	//
+	// 	P3[19:0] = denom
+	//
+
+	// Set PLL config registers
+	var p1, p2, p3 uint32
+	if num == 0 {
+		// Integer mode
+		p1 = 128*uint32(mult) - 512
+		p2 = num
+		p3 = denom
+	} else {
+		// Fractional mode
+		p1 = uint32(128*float64(mult) + math.Floor(128*(float64(num)/float64(denom))) - 512)
+		p2 = uint32(128*float64(num) - float64(denom)*math.Floor(128*(float64(num)/float64(denom))))
+		p3 = denom
+	}
+
+	// Get the appropriate starting point for the PLL registers
+	baseaddr := uint8(26)
+	if pll == PLL_B {
+		baseaddr = 34
+	}
+
+	// The datasheet is a nightmare of typos and inconsistencies here!
+	data := d.buf[:8]
+	data[0] = uint8((p3 & 0x0000FF00) >> 8)
+	data[1] = uint8(p3 & 0x000000FF)
+	data[2] = uint8((p1 & 0x00030000) >> 16)
+	data[3] = uint8((p1 & 0x0000FF00) >> 8)
+	data[4] = uint8(p1 & 0x000000FF)
+	data[5] = uint8(((p3 & 0x000F0000) >> 12) | ((p2 & 0x000F0000) >> 16))
+	data[6] = uint8((p2 & 0x0000FF00) >> 8)
+	data[7] = uint8(p2 & 0x000000FF)
+	d.bus.WriteRegister(d.Address, baseaddr, data)
+
+	// Reset both PLLs
+	data = d.buf[:1]
+	data[0] = (1 << 7) | (1 << 5)
+	d.bus.WriteRegister(d.Address, PLL_RESET, data)
```

can this error? Do you want to return an error if it does?